app/models/pipeline_run.rb

```diff
@@ -160,31 +166,168 @@ def succeeded?
+      end
+    end
+
+    job_stats.delete_all
+    self.job_stats_attributes = stats_array
+    self.taxon_counts_attributes = taxon_counts_attributes_filtered
+    self.updated_at = Time.now.utc
+    save
+    # aggregate the data at genus level
+    generate_aggregate_counts('genus') unless pr.multihit?
+    # merge more accurate name information from lineages table
+    update_names
+    # denormalize superkingdom_taxid into taxon_counts
+    if multihit?
+      update_superkingdoms
+    else
+      update_genera
+    end
+    # label taxa as phage or non-phage
+    update_is_phage
+
+    # rm the json
+    _stdout, _stderr, _status = Open3.capture3("rm -f #{downloaded_json_path} #{downloaded_stats_path}")
+  end
+
+  def db_load_postprocess
+    byteranges_json_s3_path = "#{postprocess_output_s3_path}/#{TAXID_BYTERANGE_JSON_NAME}"
+    downloaded_byteranges_path = PipelineRun.download_file(byteranges_json_s3_path, local_json_path)
+    taxon_byteranges_csv_file = "#{local_json_path}/taxon_byteranges"
+    hash_array_json2csv(downloaded_byteranges_path, taxon_byteranges_csv_file, %w[taxid hit_type first_byte last_byte])
+    ` cd #{local_json_path};
+      sed -e 's/$/,#{id}/' -i taxon_byteranges;
+      mysqlimport --replace --local --user=$DB_USERNAME --host=#{rds_host} --password=$DB_PASSWORD --columns=taxid,hit_type,first_byte,last_byte,pipeline_run_id --fields-terminated-by=',' idseq_#{Rails.env} taxon_byteranges;
+    `
+    _stdout, _stderr, _status = Open3.capture3("rm -f #{downloaded_byteranges_path}")
+  end
+
+  def s3_output_for(db_load_command)
+    case db_load_command
+    when "db_load_host_filtering"
+      "#{host_filter_output_s3_path}/#{STATS_JSON_NAME}"
+    when "db_load_alignment"
+      "#{alignment_output_s3_path}/#{STATS_JSON_NAME}"
+    when "db_load_postprocess"
+      "#{postprocess_output_s3_path}/#{TAXID_BYTERANGE_JSON_NAME}"
+    end
+  end
+
+  def output_ready?(db_load_command)
+    file_generated_since_run(s3_output_for(db_load_command))
+  end
+
+  def monitor_results
+    return if completed?
+    update(pipeline_version: fetch_pipeline_version) if pipeline_version.blank?
+    if output_ready?("db_load_host_filtering") && ![LOADING_QUEUED, HOST_FILTER_LOADED].include?(result_status)
+      update(result_status: LOADING_QUEUED)
+      Resque.enqueue(ResultMonitorLoad, id, "db_load_host_filtering", HOST_FILTER_LOADED)
+    end
+    if output_ready?("db_load_alignment") && ![LOADING_QUEUED, ALIGNMENT_LOADED].include?(result_status)
+      update(result_status: LOADING_QUEUED)
+      Resque.enqueue(ResultMonitorLoad, id, "db_load_alignment", ALIGNMENT_LOADED)
+    end
+    if output_ready?("db_load_postprocess") && ![LOADING_QUEUED, POSTPROCESS_LOADED].include?(result_status)
+      update(result_status: LOADING_QUEUED)
+      Resque.enqueue(ResultMonitorLoad, id, "db_load_postprocess", POSTPROCESS_LOADED)
+    end
+    if job_status == POSTPROCESS_LOADED # last output has been loaded
```

Comment:
I think 305-315 shouldn't be here actually. job_status and result_status should be separate. 
the finalize flag should only be set by job_monitor. 

I think you can do 
if result_status == POSTPROCESS_LOADED 
self.update(result_status: STATUS_CHECKED) 
end

Reply:
ok, I changed it so job_status is not here and only belongs to pipeline monitor. 

I've also added a `results_complete?` function and I think this should be used for anything to do with the web app. Web cares only about whether results are present, not about what happened with the job execution. For this reason, I left `notify_users` and `create_or_update_project_background` here in `monitor_results`, now conditional on `results_complete?` rather than `complete?` (=`job_complete?`).